### PR TITLE
Implement Clean Label Backdoor Attacks

### LINKED ISSUE #666: Add clean label backdoor attack
**Is your feature request related to a problem? Please describe.**
Implement a new attack "clean label backdoor attack" that incorporates  a backdoor without modifying the sample's label. 

**Describe the solution you'd like**
https://people.csail.mit.edu/madry/lab/cleanlabel.pdf 

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by ebubae**
# Description

Implements Clean Label Backdoor Attack by Turner et. al. 2018. 

Fixes #666 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Added modern ART tests for CLBD attack.

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/attacks/poison/test_clean_label_backdoor_attack.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+@pytest.mark.skipMlFramework("non_dl_frameworks", "pytorch", "mxnet")
+def test_failure_modes(art_warning, get_default_mnist_subset, image_dl_estimator, params):
+    try:
+        (x_train, y_train), (_, _) = get_default_mnist_subset
+        classifier, _ = image_dl_estimator()
+        target = to_categorical([9], 10)[0]
+        backdoor = PoisoningAttackBackdoor(add_pattern_bd)
+        attack = PoisoningAttackCleanLabelBackdoor(backdoor, classifier, target, **params)
+    except ARTTestException as e:
+        art_warning(e)
```
Comment (beat-buesser) [inline tests/attacks/poison/test_clean_label_backdoor_attack.py:60]: Why is this test expected to fail?

File: art/attacks/poisoning/clean_label_backdoor_attack.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+import logging
+import numpy as np
+
+from typing import Optional, Tuple, TYPE_CHECKING, Union
+
+from art.attacks import PoisoningAttackBlackBox
+from art.attacks.evasion import ProjectedGradientDescent
+from art.attacks.poisoning import PoisoningAttackBackdoor
+from art.estimators.classification.classifier import ClassifierLossGradients
```
Comment (beat-buesser) [inline art/attacks/poisoning/clean_label_backdoor_attack.py:31]: It looks like this import is unused, can we remove it?

File: art/utils.py [hunk 3]
```diff
@@ -391,7 +391,7 @@ def tanh_to_original(
 # --------------------------------------------------------------------------------------------------- LABELS OPERATIONS
 
 
-def to_categorical(labels: np.ndarray, nb_classes: Optional[int] = None) -> np.ndarray:
+def to_categorical(labels: Union[np.ndarray, List[float]], nb_classes: Optional[int] = None) -> np.ndarray:
```
Comment (beat-buesser) [inline art/utils.py:394]: Why is the list of floats type required? Could we instead convert potential list input into a ndarray?

File: tests/attacks/poison/test_clean_label_backdoor_attack.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+from art.attacks.poisoning import PoisoningAttackCleanLabelBackdoor, PoisoningAttackBackdoor
+from art.attacks.poisoning.perturbations import add_pattern_bd
+from art.utils import to_categorical
+
+from tests.utils import ARTTestException
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.skipMlFramework("non_dl_frameworks", "pytorch", "mxnet")
```
Comment (beat-buesser) [inline tests/attacks/poison/test_clean_label_backdoor_attack.py:33]: Why are the tests skipping pytorch and mxnet?

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @ebubae Thank you for implementing Clean Label Backdoor attacks! I only have a few minor review comments and questions above.

Could you please run the Black code formatter on all changed files with option `--line-length 120`?
Would it be possible to hide the TensorFlow warnings in cell 8 of the new notebook? If not it's ok too.

File: tests/attacks/poison/test_clean_label_backdoor_attack.py (hunk 1 above)
Comment (ebubae) [reply tests/attacks/poison/test_clean_label_backdoor_attack.py:60]: That test is checking for invalid configurations of the attack. In this case, `params` is one of a set of invalid hyper-parameters (wrongly sized arrays, negative percentages, etc). I use this to check that `check_params` is working as expected.

File: art/utils.py (hunk 3 above)
Comment (ebubae) [reply art/utils.py:394]: The conversion happens in the first line of the function, so the function already works just as well for lists of floats and np arrays. I was just adding the type, so lists could also be accepted.

File: tests/attacks/poison/test_clean_label_backdoor_attack.py (hunk 4 above)
Comment (ebubae) [reply tests/attacks/poison/test_clean_label_backdoor_attack.py:33]: mxnet and pytorch both had some errors with the backdoor perturbation. I assume this has something to do with the shape of backdoor inputs. I've opened an issue to investigate this in #745

File: tests/attacks/poison/test_clean_label_backdoor_attack.py (hunk 1 above)
Comment (ebubae) [reply tests/attacks/poison/test_clean_label_backdoor_attack.py]: I've changed this test to now expect a `ValueError` instead of a failure.

**Review by beat-buesser — APPROVED** [review]
